lib/retry.js

```diff
@@ -0,0 +1,175 @@
+      timeoutFactor: timeoutFactor ?? 2,
+      max: max ?? 8,
+      //   Indicates weather or not retry on methods
+      //  Takes prevalence over idempotent
+      methods: methods ?? [
+        'GET',
+        'HEAD',
+        'OPTIONS',
+        'PUT',
+        'DELETE',
+        'TRACE',
+        'PATCH'
+      ],
+      //   States weather or not retry on idempotent methods
+      idempotent: idempotent ?? true,
+      // Works as block list if status is not provided
+      status: statusCode ?? [420, 429, 502, 503, 504],
+      //   List of errors to not retry
+      codes: codes ?? [
+        'ECONNRESET',
+        'ECONNREFUSED',
+        'ENOTFOUND',
+        'ENETDOWN',
+        'ENETUNREACH',
+        'EHOSTDOWN',
+        'EHOSTUNREACH',
+        'EPIPE'
+      ]
+    }
+
+    // TODO: hidde behind a symbol
+    this.count = 0
+    this.timeout = null
+    this.retryAfter = this.retryOpts.timeout
+    this.range = opts.method === 'GET' ? parseRange(opts.headers?.range) : null
+  }
+
+  onConnect (abort) {
+    this.abort = abort
+    return this.handler.onConnect(reason => {
+      this.aborted = true
+      this.abort(reason)
+    })
+  }
+
+  onBodySent (chunk) {
+    return this.handler.onBodySent(chunk)
+  }
+
+  //   TODO: hidde behind a symbol
+  retry (_err, { counter, currentTimeout }, opts) {
+    const { max, maxTimeout, timeoutFactor } = opts.retry
+
+    if (counter > max) return null
+
+    const retryTimeout = Math.min(
+      currentTimeout * timeoutFactor ** counter,
+      maxTimeout
+    )
+
+    return retryTimeout
+  }
+
+  onHeaders (statusCode, rawHeaders, resume, statusMessage) {
+    if (statusCode < 400 || this.opts.retry.status.includes(statusCode)) {
+      return this.handler.onHeaders(
+        statusCode,
+        rawHeaders,
+        resume,
+        statusMessage
+      )
+    }
+
+    const err = createError(statusCode, { headers: parseHeaders(rawHeaders) })
+
+    // Meant to just indicate weather or not trigger a retry
+    const retryAfter = this.retryOpts.retry(
+      err,
+      { counter: this.count++, currentTimeout: this.retryAfter },
+      this.opts
+    )
+
+    if (retryAfter == null) {
+      return this.handler.onHeaders(
+        statusCode,
+        rawHeaders,
+        resume,
+        statusMessage
+      )
+    }
+
+    assert(Number.isFinite(retryAfter), 'invalid retryAfter')
+
+    this.retryAfter = retryAfter
+
+    this.abort(err)
+
+    return false
+  }
+
+  onData (chunk) {
+    return this.handler.onData(chunk)
+  }
+
+  onComplete (rawTrailers) {
+    return this.handler.onComplete(rawTrailers)
+  }
+
+  onError (err) {
+    if (this.timeout) {
+      clearTimeout(this.timeout)
+      this.timeout = null
+    }
+
+    // TODO: improve the erro handler for retry
+    if (
+      this.retryAfter == null ||
+      this.aborted ||
+      isDisturbed(this.opts.body)
+    ) {
+      return this.handler.onError(err)
+    }
+
+    this.timeout = setTimeout(() => {
```

Comment:
This is only valid if we errored before forwarding any headers or body,

Reply:
Or failed while connecting to the origin